### PR TITLE
Transformer{DecoderLayer} : no batch dim

### LINKED ISSUE #60585: Rollup: No-batch-dim support for torch.nn modules
## Background
A previous issue (https://github.com/pytorch/pytorch/issues/47149) requested support of arbitrary batch dimensions across module inputs. This rollup issue details the `torch.nn` module updates required to address a subset of this functionality: specifically, the case of no batch dimensions. This particular case is useful for composability with a future vmap.

## Module support
- [x] `nn.AdaptiveAvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveAvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveAvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveLogSoftmaxWithLoss` - Should be updated to support 1D input and scalar target / output1 (@george-qi)
- [x] `nn.AdaptiveMaxPool1d` - Should be updated to support 2D inputs
- [x] `nn.AdaptiveMaxPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AdaptiveMaxPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.AlphaDropout`
- [x] `nn.AvgPool1d` - Should be updated to support 2D inputs
- [x] `nn.AvgPool2d` - Supports 3D inputs (no batch), but docs should make this clear
- [x] `nn.AvgPool3d` - Supports 4D inputs (no batch), but docs should make this clear
- [x] `nn.BCELoss`
- [x] `nn.BCEWithLogitsLoss`
- [x] `nn.Bilinear` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.CELU`
- [x] `nn.CTCLoss` - Should be updated to support a single (source, target) sequence (@george-qi)
- [x] `nn.ConstantPad1d`
- [x] `nn.ConstantPad2d`
- [x] `nn.ConstantPad3d`
- [x] `nn.Conv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.Conv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.Conv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.ConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.ConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.ConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.CosineEmbeddingLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.CosineSimilarity` (pass `dim=0` for no-batch-dim support)
- [x] `nn.CrossEntropyLoss` - Should be updated to support 1D input with scalar target (@kshitij12345 )
- [x] `nn.Dropout`
- [x] `nn.Dropout2d`
- [x] `nn.Dropout3d`
- [x] `nn.ELU`
- [x] `nn.Embedding`
- [x] `nn.FeatureAlphaDropout` - Already supports no batch dims, but docs are missing
- [x] `nn.Flatten` (pass `start_dim=0` for no-batch-dim support)
- [x] `nn.Fold` - Should be updated to support 3D inputs (@kshitij12345)
- [x] `nn.FractionalMaxPool2d` - Should be updated to support 3D inputs
- [x] `nn.FractionalMaxPool3d` - Should be updated to support 4D inputs (@george-qi)
- [x] `nn.GELU`
- [x] `nn.GLU`
- [x] `nn.GRU` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.GRUCell` - Should be updated to support 1D input / hidden state (@kshitij12345)
- [x] `nn.GaussianNLLLoss` - Already supports no batch dims, but docs should make this clear (@george-qi)
- [x] `nn.Hardshrink`
- [x] `nn.Hardsigmoid`
- [x] `nn.Hardswish`
- [x] `nn.Hardtanh`
- [x] `nn.HingeEmbeddingLoss`
- [x] `nn.HuberLoss`
- [x] `nn.Identity`
- [x] `nn.InstanceNorm1d` - Should be updated to support 1D or 2D inputs (@kshitij12345)
- [x] `nn.InstanceNorm2d` - Should be updated to support 2D or 3D inputs (@kshitij12345)
- [x] `nn.InstanceNorm3d` - Should be updated to support 3D or 4D inputs (@kshitij12345)
- [x] `nn.KLDivLoss`
- [x] `nn.L1Loss`
- [x] `nn.LPPool1d` - Should be updated to support 2D inputs
- [x] `nn.LPPool2d`
- [x] `nn.LSTM` - Should be updated to support single input, hidden state, cell state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.LSTMCell` - Should be updated to support 1D input, hidden state, cell state (@kshitij12345)
- [x] `nn.LayerNorm` (pass correctly-set `normalized_shape` for no-batch-dim support)
- [x] `nn.LazyConv1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConv2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConv3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyConvTranspose1d` - Should be updated to support 2D input (@jbschlosser )
- [x] `nn.LazyConvTranspose2d` - Should be updated to support 3D input (@jbschlosser )
- [x] `nn.LazyConvTranspose3d` - Should be updated to support 4D input (@jbschlosser )
- [x] `nn.LazyLinear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LeakyReLU`
- [x] `nn.Linear` - Already supports no batch dims, but docs should make this clear
- [x] `nn.LogSigmoid`
- [x] `nn.LogSoftmax`
- [x] `nn.MSELoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MarginRankingLoss` - Should be updated to support scalars (@kshitij12345)
- [x] `nn.MaxPool1d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool2d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxPool3d` - Already supports no batch dims when `return_indices=False`, but need to support `return_indices=True`
- [x] `nn.MaxUnpool1d` - Should be updated to support 2D input / indices
- [x] `nn.MaxUnpool2d` - Should be updated to support 3D input / indices
- [x] `nn.MaxUnpool3d` - Should be updated to support 4D input / indices
- [x] `nn.Mish`
- [x] `nn.MultiLabelMarginLoss`
- [x] `nn.MultiLabelSoftMarginLoss` - Should be updated to support 1D input / target
- [x] `nn.MultiMarginLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.MultiheadAttention` - Should be updated to support unbatched query, key, values, and masks; `batch_first` arg is meaningless in this case
- [x] `nn.NLLLoss` - Should be updated to support 1D input and scalar target
- [x] `nn.PairwiseDistance` - Should be updated to support 1D inputs with scalar output
- [x] `nn.PixelShuffle`
- [x] `nn.PixelUnshuffle`
- [x] `nn.PoissonNLLLoss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.RNN` - Should be updated to support a single input sequence / hidden state; `batch_first` arg is meaningless in this case (@kshitij12345)
- [x] `nn.RNNCell` - Should be updated to support 1D input and hidden state (@kshitij12345)
- [x] `nn.RReLU`
- [x] `nn.ReLU`
- [x] `nn.ReLU6`
- [x] `nn.ReflectionPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReflectionPad2d`
- [x] `nn.ReflectionPad3d`
- [x] `nn.ReplicationPad1d` - Should be updated to support 2D inputs
- [x] `nn.ReplicationPad2d`
- [x] `nn.ReplicationPad3d`
- [x] `nn.SELU`
- [x] `nn.SiLU`
- [x] `nn.Sigmoid`
- [x] `nn.SmoothL1Loss` - Already supports no batch dims, but docs should make this clear
- [x] `nn.SoftMarginLoss`
- [x] `nn.Softmax`
- [x] `nn.Softmax2d` - Should be updated to support 3D input
- [x] `nn.Softmin`
- [x] `nn.Softplus`
- [x] `nn.Softshrink`
- [x] `nn.Softsign`
- [x] `nn.Tanh`
- [x] `nn.Tanhshrink`
- [x] `nn.Threshold`
- [x] `nn.Transformer` - Should be updated to support unbatched source / target sequences and masks (@kshitij12345)
- [x] `nn.TransformerDecoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TransformerEncoderLayer` - Should be updated to support unbatched inputs (@kshitij12345)
- [x] `nn.TripletMarginLoss` - Should be updated to support 1D anchor, positive, and negative (@kshitij12345)
- [x] `nn.TripletMarginWithDistanceLoss` (support is based on the choice of distance function)
- [x] `nn.Unflatten` - Already supports no batch dims, but docs should make this clear
- [x] `nn.ZeroPad2d`

## Semantically incompatible modules
* `nn.BatchNorm1d` - Only defined over batch
* `nn.BatchNorm2d` - Only defined over batch
* `nn.BatchNorm3d` - Only defined over batch
* `nn.LazyBatchNorm1d` - Only defined over batch
* `nn.LazyBatchNorm2d` - Only defined over batch
* `nn.LazyBatchNorm3d` - Only defined over batch
* `nn.SyncBatchNorm` - Only defined over batch

## Modules that require BC-breaking changes
* `nn.ChannelShuffle` - While the docs say that the module accepts shape (*, C, H, W), the implementation assumes shape (N, C, *) with * being 1 or more dimensions; switching to (C, *) is BC-breaking because it would reinterpret dims
* `nn.EmbeddingBag` - Already supports 1D inputs with different semantics
* `nn.GroupNorm` - Supports arbitrary spatial dims; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.LocalResponseNorm` - Supports (N, C, *) shape now; switching to (C, *) is BC-breaking since it would reinterpret dims
* `nn.PReLU` - Always assumes the 2nd dim is the channels dim
* `nn.Unfold` - Supports 4D inputs; switching to (C, *) instead of (N, C, *) is BC-breaking since it would reinterpret dims
* `nn.Upsample` - Supporting unbatched inputs would be BC-breaking because it would reinterpret dims

## Irrelevant modules
* `nn.Module`
* `nn.ModuleDict`
* `nn.ModuleList`
* `nn.ParameterDict`
* `nn.ParameterList`
* `nn.Sequential`
* `nn.TransformerDecoder` - see `nn.TransformerDecoderLayer` instead
* `nn.TransformerEncoder` - see `nn.TransformerEncoderLayer` instead

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #60585 

TransformerDecoder Test Timings (takes about 30s)
<details>

```
pytest test/test_modules.py -k _TransformerDeco --durations=10
============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.0, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /home/kshiteej/Pytorch/pytorch_no_batch_mha, configfile: pytest.ini
plugins: hypothesis-6.23.2, repeat-0.9.1
collected 639 items / 591 deselected / 48 selected                                                                                                                                                               

test/test_modules.py ss......ss......ss..ssssssssss..................                                                                                                                                      [100%]

================================================================================================================================================================================ slowest 10 durations ==============================================================================================
17.13s call     test/test_modules.py::TestModuleCUDA::test_gradgrad_nn_TransformerDecoderLayer_cuda_float64
4.13s call     test/test_modules.py::TestModuleCPU::test_gradgrad_nn_TransformerDecoderLayer_cpu_float64
1.22s call     test/test_modules.py::TestModuleCUDA::test_grad_nn_TransformerDecoderLayer_cuda_float64
0.86s call     test/test_modules.py::TestModuleCPU::test_cpu_gpu_parity_nn_TransformerDecoderLayer_cpu_float32
0.73s call     test/test_modules.py::TestModuleCUDA::test_cpu_gpu_parity_nn_TransformerDecoderLayer_cuda_float32
0.57s call     test/test_modules.py::TestModuleCUDA::test_non_contiguous_tensors_nn_TransformerDecoderLayer_cuda_float32
0.56s call     test/test_modules.py::TestModuleCUDA::test_non_contiguous_tensors_nn_TransformerDecoderLayer_cuda_float64
0.48s call     test/test_modules.py::TestModuleCPU::test_grad_nn_TransformerDecoderLayer_cpu_float64
0.41s call     test/test_modules.py::TestModuleCUDA::test_multiple_device_transfer_nn_TransformerDecoderLayer_cuda_float32
0.40s call     test/test_modules.py::TestModuleCUDA::test_cpu_gpu_parity_nn_TransformerDecoderLayer_cuda_float64
============================================================================================ short test summary info =============================================================================================
========================================================================== 32 passed, 16 skipped, 591 deselected, 3 warnings in 29.62s ===========================================================================
```

</details>

Transformer Test Timings (takes about 1m10s)

<details>
```
pytest test/test_modules.py -k _Transformer_ --durations=10
============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.0, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /home/kshiteej/Pytorch/pytorch_no_batch_mha, configfile: pytest.ini
plugins: hypothesis-6.23.2, repeat-0.9.1
collected 639 items / 591 deselected / 48 selected                                                                                                                                                               

test/test_modules.py ss......ss......ss..ssssssssss..................                                                                                                                                      [100%]

==================================================================================
============================================================================================== slowest 10 durations ==============================================================================================
46.40s call     test/test_modules.py::TestModuleCUDA::test_gradgrad_nn_Transformer_cuda_float64
11.09s call     test/test_modules.py::TestModuleCPU::test_gradgrad_nn_Transformer_cpu_float64
2.48s call     test/test_modules.py::TestModuleCUDA::test_grad_nn_Transformer_cuda_float64
1.03s call     test/test_modules.py::TestModuleCPU::test_grad_nn_Transformer_cpu_float64
0.96s call     test/test_modules.py::TestModuleCUDA::test_cpu_gpu_parity_nn_Transformer_cuda_float32
0.87s call     test/test_modules.py::TestModuleCUDA::test_non_contiguous_tensors_nn_Transformer_cuda_float32
0.85s call     test/test_modules.py::TestModuleCUDA::test_non_contiguous_tensors_nn_Transformer_cuda_float64
0.85s call     test/test_modules.py::TestModuleCPU::test_cpu_gpu_parity_nn_Transformer_cpu_float32
0.65s call     test/test_modules.py::TestModuleCUDA::test_cpu_gpu_parity_nn_Transformer_cuda_float64
0.47s call     test/test_modules.py::TestModuleCUDA::test_multiple_device_transfer_nn_Transformer_cuda_float32
============================================================================================ short test summary info =============================================================================================
===================================================================== 32 passed, 16 skipped, 591 deselected, 3 warnings in 70.19s (0:01:10) ======================================================================
```
</details>

### CODE REVIEW

**Review by kshitij12345 — COMMENTED** [review]
I thought of using the same sample_fn for TransformerEncoder and Decoder (but decoder takes a mask as well), ended up keeping them seperate.

**Review by jbschlosser — APPROVED** [review]
LGTM! thanks :)
